### ISSUE TITLE
Adding field group name as a parameter to dcgm sampler

### ISSUE THREAD

**Issue by ronnieChatt** [LLNL/ldms-plugins-llnl#13]
Currently the dcgm sampler creates a `dcgmfieldgroup` with the name `ldmsd_fields`. 

https://github.com/LLNL/ldms-plugins-llnl/blob/f9f5fa9a583b7f9274a2ee83d148b7a8f30ae181/src/dcgm.c#L193-L194

This creates a field group with that name, and we can see that below: 

> arghyac@cgpu11:~/ldms_ronnie.archive> dcgmi fieldgroup -l
...
...
...
...
...
+-------------------+----------------------------------------------------------+
| FIELD GROUPS                                                                 |
+===================+==========================================================+
| ID                | 4                                                        |
| Name              | ldmsd_fields                                             |
| Field IDs         | 150, 155, 203                                            |
+-------------------+----------------------------------------------------------+

This restricts users (for testing purposes) to run another `ldms` daemon and connect with the `dcgm` producer because that creates an *duplicate key passed error*. 

> Fri May 28 14:54:17 2021: DEBUG     : Checking for dcgm in /global/homes/a/arghyac/ldms_ronnie.archive/ovis-ldms-4.3.7/lib/ovis-ldms/
Fri May 28 14:54:17 2021: DEBUG     : dcgm get_plugin() called (ovis-ldms 4.3.7)
Fri May 28 14:54:17 2021: DEBUG     : msg_no 3 flags 3 rec_len 28 rsp_err 0
Fri May 28 14:54:17 2021: DEBUG     : dcgm config() called
Fri May 28 14:54:17 2021: ERROR     : dcgm dcgmFieldGroupCreate failed: Duplicate Key passed to function(-26)
Fri May 28 14:54:17 2021: DEBUG     : msg_no 4 flags 3 rec_len 75 rsp_err -1
Fri May 28 14:54:17 2021: ERROR     : msg_no 4: error -1: Plugin 'dcgm' configuration error.

My guess would be because it's looking for a unique name as defined in the function def here: 

https://github.com/NVIDIA/gpu-monitoring-tools/blob/master/bindings/go/dcgm/dcgm_agent.h#L553-L554

Why is this needed ? 

We (at NERSC) are testing this feature / support for our GPU systems and the developers / interns / staff involved in the project do not have the proper rights to connect to the root enabled ldms. 

Would there be a workaround ? Can we:
1. manually edit the `dcgm.c` src on a local installation and add a different name to the field group --> **edit: Managed to get this working**, but still would be interested to know if there are alternatives to this approach.  
2. can we pass some `field group` identifier to the sampler.conf file. Provided the user is creating their own group: 
Eg: `dcgmi fieldgroup -c ronnie_ldms -f 150, 155` and then passing something like `fieldgroups = ronnie_ldms` ... fields = 150,155` in the sampler.conf. 
3. There might be other ways that I am not thinking of...

Any help on this would be really appreciated. 

Thanks, 
Ronnie Chatterjee
Applications Performance Group, NERSC

**Comment by morrone** [comment]
I think that just adding the process ID to the string would be sufficient to eliminate conflicts.

You must keep in mind though that _all_ of the users of DCGM must use the sample sampling rate. If they don't, the last one to request a sampling rate wins, and that rate effects _all_ users of DCGM without them knowing the rate has been changed.

**Comment by ronnieChatt** [comment]
Yeah by changing the `fieldgroup` name and recompiling I could make it work without any issues. Yeah I think the process id + string makes more sense. 

Thanks for the heads up on the DCGM sampling rate, that's good to know. 

But say if you have a `ldmsd` already running on the system with a `dcgm` producer before a new ldms plugin with dcgm is initiated (new user) would that affect the initial (root enabled in our case) sampling rate ?

**Comment by morrone** [comment]
> But say if you have a `ldmsd` already running on the system with a `dcgm` producer before a new ldms plugin with dcgm is initiated (new user) would that affect the initial (root enabled in our case) sampling rate ?

Yes, it will. At least if the new dcgm is looking at the same fields...some sampling rates may be per field. My memory is not clear on that.

**Comment by ronnieChatt** [comment]
Thanks @morrone. We will keep that in mind. 

One additional thing, just like the other `ldms` producers, does `ldms-dcgm` plugin let's us capture the UUID of the GPU, so in cases if you have 1 node with 8 GPUs, and your `nv-hostengine` is running as root, `dcgm` records all the GPUs on that node with GPU_IDs --> 0-n. Is there a way to add a parameter in the `dcgm` producer in the sampler or somewhere to capture the UUID for cases: 
1. where `not` all the GPUs were requested by a user (shared usage of a node) ; 
1. where even if the user asked for exclusive access they're only running on 2 GPUs out of n on that node. 

I don't see any SchedMD parameters to connect the dcgm with, wanted to pick your brain if this is something you already have some support for or ways you can think this can be achieved.

**Comment by morrone** [comment]
I guess I'm not seeing the connection between multiple users sharing a node and the use and UUIDs. Why wouldn't LDMS monitor all the GPUs regardless of how many users there are? And why would adding a UUID improve the situation over the integer identifiers for the GPUs?
